lib/SILOptimizer/Transforms/CopyPropagation.cpp

```diff
@@ -0,0 +1,908 @@
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// SSA Copy propagation pass to remove unnecessary copy_value and destroy_value
+/// instructions.
+///
+/// Because this algorithm rewrites copies and destroys without attempting
+/// to balance the retain count, it is only sound for ownership-SSA.
+///
+/// [WIP] This complements -enable-sil-opaque-values. It is not yet enabled by
+/// default. It can be enabled once the ownership properties, currently
+/// specified as helpers at the top of this file, are formalized via a central
+/// API.
+///
+/// The pass has four steps:
+///
+/// 1. Find SSA defs that have been copied.
+///
+/// Then, separately, for each copied def, perform steps 2-4:
+///
+/// 2. Compute "pruned" liveness of the def and its copies, ignoring original
+///    destroys.
+///
+/// 3. Find the SSA def's final destroy points based on its pruned liveness.
+///
+/// 4. Rewrite the def's original copies and destroys, inserting new copies
+/// where needed.
+///
+/// The state used by this pass is encapsulated in CopyPropagationState, simply
+/// refered to as `pass`:
+///
+/// - Step 1 initializes `pass.currDef`.
+///
+/// - Step 2 initializes `pass.liveness`.
+///
+/// - Step 3 initializes `pass.destroys` and inserts destroy_value instructions.
+///
+/// - Step 4 deletes original copies and destroys and inserts new copies.
+///
+/// Example #1: Handle consuming and nonconsuming uses.
+///
+/// bb0(%arg : @owned $T, %addr : @trivial $*T):
+///   %copy = copy_value %arg : $T
+///   debug_value %copy : $T
+///   store %copy to [init] %addr : $*T
+///   debug_value %arg : $T
+///   debug_value_addr %addr : $*T
+///   destroy_value %arg : $T
+///
+/// Will be transformed to:
+///
+/// bb0(%arg : @owned $T, %addr : @trivial $*T):
+/// (The original copy is deleted.)
+///   debug_value %copy : $T
+/// (A new copy_value is inserted before the consuming store.)
+///   %copy = copy_value %0 : $T
+///   store %3 to [init] %1 : $*T
+/// (The non-consuming use now uses the original value.)
+///   debug_value %0 : $T
+/// (A new destroy is inserted after the last use.)
+///   destroy_value %0 : $T
+///   debug_value_addr %1 : $*T
+/// (The original destroy is deleted.)
+///
+/// Example #2: Handle control flow.
+///
+/// bb0(%arg : @owned $T, %addr : @trivial $*T, %z : @trivial $Builtin.Int1):
+///   cond_br %z, bb1, bb2
+///
+/// bb1:
+///   debug_value %arg : $T
+///   %copy = copy_value %arg : $T
+///   destroy_value %copy : $T
+///   br bb2
+///
+/// bb2:
+///   destroy_value %arg : $T
+///
+/// Will be transformed to:
+///
+///  bb0(%0 : @owned $T, %1 : @trivial $*T, %2 : @trivial $Builtin.Int1):
```

Comment:
s/%0/%arg/
(so that it's easier to correlate with the original SIL)
also in the following lines